rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,638 @@
+dnnl_status_t dnnl_brgemm_execute(const_dnnl_brgemm_t brg,
+        const void *A, const void *B, const void **A_B_offsets,
+        void *C_ptr, void *scratchpad_ptr,
+        const_ukernel_attr_params_t attr_arguments);
+
+dnnl_status_t dnnl_brgemm_execute_with_postops(const_dnnl_brgemm_t brg,
+        const void *A, const void *B, const void **A_B_offsets,
+        const void *C_ptr, void *D_ptr, void *scratchpad_ptr,
+        const_ukernel_attr_params_t attr_arguments);
+```
+
+### Simple example for bf16 matmul with f32 accumulation and relu fusion.#include "dnnl_block.h"
+
+```c++
+void matmul_with_relu(const void *src, const void *weights, void *dst) {
+  // Here we will create a bf16 matmul operation, with f32 accumulation,
+  // and downconversion to bf16 at the end
+  // We assume a copy based implementation: we copy just B for example simplification
+  // For the D matrix, we will write to destination directly.
+
+    // we assume divisibility to simplify example
+    // BK is here just for the purpose of example, in practice it should be much larger.
+    dim_t BM = 16, BN = 16, BK = 32;
+    assert(M % BM == 0);
+    assert(N % BN == 0);
+    assert(K % BK == 0);
+
+    brgemm brg(K / BK, BM, BN, BK,
+            data_type::bf16, K,  // type/ld for A/src
+            data_type::bf16, BN, // type/ld for B/weights
+            data_type::f32, BN); // type/ld for C, temporary buffer
+    
+    // Apply post-op and downconversion to bf16
+    post_ops po;
+    po.append_eltwise(algorithm::eltwise_relu, 0.f, 0.f);
+    brg.set_post_ops(po, data_type::bf16, N); // type/ld for D/dst)
+    brg.finalize();
+
+    assert(brg.get_A_pack_type() == pack_type::plain);
+    transform tB(K, BN, bf16, pack_type::plain, N, brg.get_B_pack_type(), BN);
+    tB.finalize();
+
+    // Here we jit the block level kernels
+    brg.generate();
+    tB.generate();
+
+    // we do parallelisation based on M/N slicing
+    // each thread gets a single output block
+    parallel_for(range(M/BM, N/BN),
+            [&](range r){
+            std::vector<std::pair> A_B;
+            A_B.reserve(K/BK);
+
+            // We allocate temporary buffers
+            // If allocator has thread-local pools, this is fine
+            void *c_buff = myalloc(BM * BN * sizeof(float));
+            void *packedB = myalloc(K * BN * sizeof(bfloat16_t));
+            void *scratch = myalloc(brg.get_scratchpad_size());
+
+            // pack B
+            tB.execute(weights + r[1] * BN * sizeof(bfloat16_t), packedB);
+
+            // we set up the arguments for the execution
+            for (int k = 0; k < K; k+=BK) {
+                A_B.emplace_back(src + (r[0] * BM * K + k) * sizeof(bfloat16_t), 
+                        packedB + k * BN * sizeof(bfloat16_t));
+            }
+
+            // we run it
+            brg.execute(A_B, c_buff, dst + (r[0] * BM * N + r[1] * BN) * sizeof(bfloat16_t), scratch);
+    });
```

Comment:
Will the original value of `c_buff` be accumulated to the final result? Is it possible to allow user to control the original value of C to participate in the accumulation. This can save the overhead of setting values of c_buf to 0.

Reply:
Yes this is controlled by the `add_C` attribute. When set to false, it will just override C with the result (same as if c_buff is 0).